policy.go

```diff
@@ -305,3 +454,230 @@ func readAndValidateLockNVIndexPublic(tpm *tpm2.TPMContext, index tpm2.ResourceC
+}
+
+// computeDynamicPolicy computes the part of an authorization policy associated with a sealed key object that can change and be
+// updated.
+func computeDynamicPolicy(alg tpm2.HashAlgorithmId, input *dynamicPolicyComputeParams) (*dynamicPolicyData, error) {
+	var pcrData []policyPCRData
+	for _, p := range input.pcrParams {
+		if len(p.digests) == 0 {
+			return nil, fmt.Errorf("no digests provided for PCR%d", p.pcr)
+		}
+		var orDigests tpm2.DigestList
+		for _, d := range p.digests {
+			trial, _ := tpm2.ComputeAuthPolicy(alg)
+			if len(pcrData) > 0 {
+				trial.PolicyOR(ensureSufficientORDigests(pcrData[len(pcrData)-1].OrDigests))
+			}
+			pcrDigest, pcrs := computePolicyPCRParams(alg, p.alg, d, p.pcr)
+			trial.PolicyPCR(pcrDigest, pcrs)
+			orDigests = append(orDigests, trial.GetDigest())
+		}
+		pcrData = append(pcrData, policyPCRData{PCR: p.pcr, Alg: p.alg, OrDigests: orDigests})
+	}
+
+	trial, _ := tpm2.ComputeAuthPolicy(alg)
+	if len(pcrData) > 0 {
+		trial.PolicyOR(ensureSufficientORDigests(pcrData[len(pcrData)-1].OrDigests))
+	}
+
+	operandB := make([]byte, 8)
+	binary.BigEndian.PutUint64(operandB, input.policyCount)
+	trial.PolicyNV(input.policyCountIndexName, operandB, 0, tpm2.OpUnsignedLE)
+
+	authorizedPolicy := trial.GetDigest()
+
+	// Create a digest to sign
+	h := input.signAlg.NewHash()
+	h.Write(authorizedPolicy)
+
+	// Sign the digest
+	sig, err := rsa.SignPSS(rand.Reader, input.key, input.signAlg.GetHash(), h.Sum(nil), &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash})
+	if err != nil {
+		return nil, xerrors.Errorf("cannot provide signature for initializing NV index: %w", err)
+	}
+
+	signature := tpm2.Signature{
+		SigAlg: tpm2.SigSchemeAlgRSAPSS,
+		Signature: tpm2.SignatureU{
+			Data: &tpm2.SignatureRSAPSS{
+				Hash: input.signAlg,
+				Sig:  tpm2.PublicKeyRSA(sig)}}}
+
+	return &dynamicPolicyData{
+		PCRData:                   pcrData,
+		PolicyCount:               input.policyCount,
+		AuthorizedPolicy:          authorizedPolicy,
+		AuthorizedPolicySignature: &signature}, nil
+}
+
+func executePolicySessionPCRAssertions(tpm *tpm2.TPMContext, session tpm2.SessionContext, input []policyPCRData) error {
+	for _, i := range input {
+		if err := tpm.PolicyPCR(session, nil, makePCRSelectionList(i.Alg, i.PCR)); err != nil {
+			return err
+		}
+		if err := tpm.PolicyOR(session, ensureSufficientORDigests(i.OrDigests)); err != nil {
+			var e *tpm2.TPMParameterError
+			if xerrors.As(err, &e) && e.Code() == tpm2.ErrorValue {
+				return xerrors.Errorf("unexpected session digest after executing TPM2_PolicyPCR assertion for PCR%d: %w", i.PCR, err)
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+// executePolicySession executes an authorization policy session using the supplied metadata. On success, the supplied policy
+// session can be used for authorization.
+func executePolicySession(tpm *tpm2.TPMContext, policySession tpm2.SessionContext, staticInput *staticPolicyData,
```

Comment:
this isn't used yet? sort of a meta comment, but we need to make sure either the snapd invoking code, or in higher level bits, that we don't pass this kind of low-level errors through too much, but always have a "cannot..." higher-level what are we trying to do, where we are in the code context information

Reply:
This is consumed by the unseal API, which will contain some handling to convert the errors we might expect from this function to better ones.

Reply:
thanks for the answer